core/src/main/java/com/amplifyframework/auth/options/AuthSignUpOptions.java

```diff
@@ -0,0 +1,128 @@
+            Map<String, String> validationData
+    ) {
+        this.userAttributes = userAttributes;
+        this.validationData = validationData;
+    }
+
+    /**
+     * Get additional user attributes which should be associated with this user on registration.
+     * @return additional user attributes which should be associated with this user on registration
+     */
+    @NonNull
+    public Map<String, String> getUserAttributes() {
+        return userAttributes;
+    }
+
+    /**
+     * Get a map of custom key/values to be sent as part of the sign up process.
+     * @return a map of custom key/values to be sent as part of the sign up process
+     */
+    @NonNull
+    public Map<String, String> getValidationData() {
+        return validationData;
+    }
+
+    @NonNull
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    @Override
+    public int hashCode() {
+        return ObjectsCompat.hash(
+                getUserAttributes(),
+                getValidationData()
+        );
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
+        } else if (obj == null || getClass() != obj.getClass()) {
+            return false;
+        } else {
+            AuthSignUpOptions authSignUpOptions = (AuthSignUpOptions) obj;
+            return ObjectsCompat.equals(getValidationData(), authSignUpOptions.getValidationData()) &&
+                    ObjectsCompat.equals(getUserAttributes(), authSignUpOptions.getUserAttributes());
+        }
+    }
+
+    @Override
+    public String toString() {
+        return new StringBuilder()
+                .append("AuthSignUpOptions { ")
+                .append("validationData: ")
+                .append(getValidationData())
+                .append(", userAttributes: ")
+                .append(getUserAttributes())
+                .append(" }")
+                .toString();
+    }
+
+    public static final class Builder {
+        private Map<String, String> userAttributes;
+        private Map<String, String> validationData;
+
+        public Builder() {
+            this.userAttributes = new HashMap<>();
+            this.validationData = new HashMap<>();
+        }
+
+        @NonNull
+        public Builder userAttributes(@NonNull Map<String, String> userAttributes) {
+            Objects.requireNonNull(userAttributes);
+            this.userAttributes.clear();
+            this.userAttributes.putAll(userAttributes);
+            return this;
+        }
+
+        @NonNull
+        public Builder validationData(@NonNull Map<String, String> validationData) {
+            Objects.requireNonNull(validationData);
+            this.validationData.clear();
+            this.validationData.putAll(validationData);
+            return this;
+        }
+
+        @NonNull
+        public AuthSignUpOptions build() {
+            return new AuthSignUpOptions(userAttributes, validationData);
```

Comment:
Somewhere around here is a good opportunity to wrap these in `Immutable.of(...)`, so that they are, int he created object.